### PR TITLE
[enh] replace requests by httpx

### PR DESCRIPTION
**Pull request by dalf**
## What does this PR do?

PR that replace requests by httpx.

`settings.yml` do not require any changes.

### implementation
* use [httpx.AsyncClient](https://www.python-httpx.org/api/#asyncclient) and [uvloop](https://github.com/MagicStack/uvloop)
* the `searx.poolrequests` API are not async.
* use `atexit` module to close HTTP clients (wait at most 3 seconds). Avoid a warning from httpx.

### proxy
* URL http:// supported
* URL https:// not supported
* URL sock4:// , socks5:// , socks5h:// : supported using [httpx-socks](https://pypi.org/project/httpx-socks/)

see `searx.poolrequests.iter_proxies`.

### sources IPs
* supported using a dedicated [transports](https://www.python-httpx.org/advanced/#mounting-transports).
	
### authentification
* see https://www.python-httpx.org/quickstart/#authentication
* `params['auth'] = (user, password)`: basic authentification
* `from httpx import DigestAuth; params['auth'] = DigestAuth(user, password)`: DIGEST authentification

### pool size
* `outgoing.pool_connections` and `outgoing.pool_maxsize` are ìmplementated, see `searx.poolrequests.LIMITS`.
	
### http2:
* disable by default
* can be enabled in settings.yml

### image proxy
* `poolrequests.stream` is an sync function that streams the response from the async httpx client.
* httpx supports sync client but
	* it requires to duplicates the specific implementation of httpx.AsyncHTTPTransport
	* the HTTP2 client may behave differently (according to the issues in encode/httpx)

### requests compatibility
* see https://www.python-httpx.org/compatibility/
* `response.url` is not a string, it is [URL](https://www.python-httpx.org/api/#url):
	* some engines have been modified to use `response.url` instead of `urlparse(response.url)`.
* `response.ok` is the `not response.is_error`.

### performance
* httpx seems faster than requests when http2 is enabled
* httpx is slowed when a lot of data are tranfered (image proxy use case).
* see https://github.com/dalf/pyhttp-benchmark

## Why is this change important?

One step to implement the new architecture.

## How to test this PR locally?

* check these engines:
   * dictzone: check the returned URL
   * elasticsearch: check authentification
   * google: check the catcha detection.
   * seznam: check the catcha detection.
   * stackoverflow: check the catcha detection.
   * yacy: check the authentification.

* check settings.yml:
    * `outgoing.request_timeout`: set a very low limit, and check the timeout
	* `outgoing.pool_maxsize`: set to 1 and send multiple request, check searx is waiting for the connection to be available
	* `outgoing.source_ips`: set multiple IP, check searx uses the different IPs.
    * `outgoing.proxies`: http, socks4, socks5, socks5h: 
		* check searx uses the proxy.
		* check searx throws error when the proxy is not available.
	
* check the image proxy (without morty) works as expected:
    * when the thumbnail is available
	* when the thumbnail is not available (HTTP error).
	* when the thumbnail is bigger than 5MB.
	
* check searx_extra/update/update_external_bangs.py fetchs the external bangs.

* check autocompletion:
	* when there is no error
	* when there is any HTTP error (no Internet connection for example).

* check everything again with http2 enable.

## Author's checklist

`tests/unit/test_poolrequests.py` needs to be updated.

## Related issues

* New version of https://github.com/searx/searx/pull/1856 (which contains general thoughts about HTTP clients, start new discussion here only about httpx).

### CODE REVIEW

File: searx/utils.py [hunk 1]
```diff
@@ -45,7 +45,7 @@ def searx_useragent():
     """Return the searx User Agent"""
     return 'searx/{searx_version} {suffix}'.format(
            searx_version=VERSION_STRING,
-           suffix=settings['outgoing'].get('useragent_suffix', ''))
+           suffix=settings['outgoing'].get('useragent_suffix', '')).strip()
```
Comment (dalf) [inline searx/utils.py:48]: a trailing white space raises a HTTP error on HTTP/1 connection (not HTTP/2).
`'searx/{searx_version} {suffix}'` --> `"searx/0.18.0 "`

File: requirements.txt [hunk 2]
```diff
@@ -8,5 +8,10 @@ lxml==4.6.2
...
 python-dateutil==2.8.1
 pyyaml==5.4.1
-requests[socks]==2.25.1
+httpx[http2]==0.17.1
+brotlipy==0.7.0
+uvloop==0.15.2; python_version >= '3.7'
+uvloop==0.14.0; python_version < '3.7'
+httpx-socks[asyncio]==0.3.1
 langdetect==1.0.8
+setproctitle==1.2.2
```
Comment (dalf) [inline requirements.txt]: Temporary dependency to help debuging: make sure there is no endless request.
see searx/webapp.py: there is a hook of `threading.Thread.__init__`

File: requirements.txt [hunk 3]
```diff
@@ -8,5 +8,10 @@ lxml==4.6.2
 pygments==2.8.0
 python-dateutil==2.8.1
 pyyaml==5.4.1
-requests[socks]==2.25.1
+httpx[http2]==0.17.1
+brotlipy==0.7.0
+uvloop==0.15.2; python_version >= '3.7'
+uvloop==0.14.0; python_version < '3.7'
```
Comment (dalf) [inline requirements.txt:14]: I'm not sure if this version should be included.
There are multiple bug fixes in the releases 0.15.* : 
* https://github.com/MagicStack/uvloop/releases

File: searx/poolrequests.py [hunk 4]
```diff
@@ -233,3 +426,96 @@ def patch(url, data=None, **kwargs):
...
+        await asyncio.gather(*[close_client(client) for client in CLIENTS.values()], return_exceptions=False)
+    future = asyncio.run_coroutine_threadsafe(close_clients(), LOOP)
+    # wait 3 seconds to close the HTTP clients
+    future.result(3)
+
+
+init()
+
+
+# ## TEMPORARY DEBUG ##
```
Comment (dalf) [inline searx/poolrequests.py]: The following lines will be removed later.
Perhaps kept in a debug plugin.

File: requirements.txt (hunk 2 above)
Comment (kvch) [reply requirements.txt]: By temporary you mean that you are removing this before the merge?

**Review by kvch — COMMENTED** [review]
I have skimmed through the PR quickly. But I would like to review it in the next few days properly.
Thanks for this! It is a great improvement! :tada: :tada: :tada:

File: searx/settings.yml [hunk 5]
```diff
@@ -67,11 +67,13 @@ ui:
...
-    request_timeout : 2.0 # default timeout in seconds, can be override by engine
+    request_timeout : 3.0 # default timeout in seconds, can be override by engine
     # max_request_timeout: 10.0 # the maximum timeout in seconds
     useragent_suffix : "" # suffix of searx_useragent, could contain informations like an email address to the administrator
-    pool_connections : 100 # Number of different hosts
-    pool_maxsize : 10 # Number of simultaneous requests by host
+    pool_connections : 100 # The maximum number of concurrent connections that may be established.
+    pool_maxsize : 20 # Allow the connection pool to maintain keep-alive connections below this point.
+    keepalive_expiry: 30.0 # Number of seconds to keep a connection in the pool
+    http2: True  # Enable HTTP/2 (experimental)
```
Comment (kvch) [inline searx/settings.yml]: If it is experimental why enable it? What is the impact of enabling this?

File: searx/settings.yml [hunk 6]
```diff
@@ -67,11 +67,13 @@ ui:
 #    key : !!binary "your_morty_proxy_key"
 
 outgoing: # communication with search engines
-    request_timeout : 2.0 # default timeout in seconds, can be override by engine
+    request_timeout : 3.0 # default timeout in seconds, can be override by engine
```
Comment (kvch) [inline searx/settings.yml]: Why is the timeout increased? Isn't this going to lead to searx returning results more slowly from users' point of view?

File: searx/webapp.py [hunk 7]
```diff
@@ -1094,6 +1120,11 @@ def config():
     })
 
 
+@app.route('/config/http')
```
Comment (kvch) [inline searx/webapp.py]: Why expose this endpoint? It does not seem useful to anyone besides developers.

File: searx/webapp.py [hunk 8]
```diff
@@ -901,50 +915,62 @@ def _is_selected_language_supported(engine, preferences):
...
-                        proxies=get_global_proxies())
+    maximum_size = 5 * 1024 * 1024
 
-    if resp.status_code == 304:
-        return '', resp.status_code
-
-    if resp.status_code != 200:
-        logger.debug('image-proxy: wrong response code: {0}'.format(resp.status_code))
-        if resp.status_code >= 400:
+    try:
```
Comment (kvch) [inline searx/webapp.py:908]: Please move out the function from the `try` that do not raise `httpx.HTTPError`. Keep the whole block as small as possible.

File: searx/poolrequests.py (hunk 4 above)
Comment (kvch) [reply searx/poolrequests.py]: Oh, I got it. The endpoint will be removed as well I assume.

File: searx/poolrequests.py [hunk 9]
```diff
@@ -152,22 +156,197 @@ def get_global_proxies():
...
+                                    verify=verify,
+                                    local_address=local_address,
+                                    max_connections=LIMITS.max_connections,
+                                    max_keepalive_connections=LIMITS.max_keepalive_connections,
+                                    keepalive_expiry=LIMITS.keepalive_expiry,
+                                    **TRANSPORT_KWARGS)
+
+
+def get_transport(verify, local_address, proxy_url):
+    global LIMITS
```
Comment (kvch) [inline searx/poolrequests.py]: Is this line necessary? AFAIK `AsyncHTTPTransport` only reads from this object.

File: searx/poolrequests.py [hunk 10]
```diff
@@ -178,24 +357,23 @@ def request(method, url, **kwargs):
...
-            raise requests.exceptions.Timeout(response=response)
-
-    # session end
-    session.close()
-
-    if hasattr(threadLocal, 'total_time'):
-        threadLocal.total_time += time_after_request - time_before_request
+    future = asyncio.run_coroutine_threadsafe(send_request(method, url, get_enable_http_protocol(), kwargs), LOOP)
+    try:
+        if timeout:
```
Comment (kvch) [inline searx/poolrequests.py]: Please move out lines that cannot raise the expectation you want to handle in this block.

**Review by kvch — CHANGES_REQUESTED** [review]
Minor issues have to be addressed. All in all, tested and works well.

File: searx/poolrequests.py (hunk 4 above)
Comment (dalf) [reply searx/poolrequests.py]: Removed

File: searx/settings.yml (hunk 6 above)
Comment (dalf) [reply searx/settings.yml]: I have forgotten to reset the value from my configuration.

Done in the last commit.

File: searx/webapp.py (hunk 7 above)
Comment (dalf) [reply searx/webapp.py]: Removed.

File: searx/webapp.py (hunk 8 above)
Comment (dalf) [reply searx/webapp.py:908]: Done in the last commit.

File: searx/poolrequests.py (hunk 10 above)
Comment (dalf) [reply searx/poolrequests.py]: Done in the last commit.

File: searx/settings.yml (hunk 5 above)
Comment (dalf) [reply searx/settings.yml]: See https://www.python-httpx.org/http2/ 

Using `AsyncHTTPTransportFixed`, I haven't seen issue on my local instance for 2 weeks.

**Comment by dalf** [comment]
Once okay, I will squash the commits.

**Comment by dalf** [comment]
Closed since the core won't evolve much in this repository according to @asciimoo

**Comment by samyak-jain** [comment]
@dalf So, is there going to be a separate repo or fork? 
Is this repo not accepting contributions or getting abondoned? Or am I misreading your comment?

**Pull request closed without merging**